### DISCUSSION TITLE
Schedule bunch a few days a week?

### DISCUSSION THREAD

**Discussion by moffat** [ttscoff/bunch-docs#38] [General]
I have some recurring events every week, and would like to have a bunch run shortly before I start and close shortly after I finish those events. It sounds like say running a bunch on mon and wed from 11-12 isn't an option at the moment, so the best option at the moment would be to create one scheduled bunch per day of the week when I need them, which then calls the bunch with the actions I need (they are the same, so I don't want to maintain two bunches). 

I could also schedule with Keyboard Maestro, etc. 

Am I missing something?

**Comment by ttscoff** [comment]
You read my mind. The next update allows things like:

```
open on: MWF 8am, T Th 9am
```

Or you can do multiple times a day with 

```
open at: 8am, 10:30am, 5pm
```

I also added a new frontmatter key for `ignores state`, and when that's set to `true` it will run the open or close commands even if the Bunch is already opened or closed (normally it would ignore a scheduled open if the Bunch was already open).

Should have the update out this week with the fun new stuff :).

**Reply by ttscoff** [reply]
Curious how this worked out for you. If you've put it to use, let me know how it's working.

**Reply by moffat** [reply]
Works great. One issue that I didn't foresee is that because I'm syncing my bunched via Dropbox (I have a laptop and desktop) the scheduled bunched will execute in the machine I'm not intending to use too. 

Any chance of including, as in keyboard maestro, a way to only execute a bunch in a particular machine. Say with a "available on:" using hostnames or something of the sort?

It isn't a big deal either way, but I could imagine I'm not the only one with multiple machines and synced bunched...

**Reply by ttscoff** [reply]
Hoo boy, yeah, that seems like a valid concern and one that I'll have to think about how to tackle. What do you think would be the easiest way to identify a machine accurately in a plain text file? hostname maybe, but on two different networks it's feasible that two machines use the same hostname. MAC Address, maybe... 

One possibility might be to have a `schedule if:` key that let you specify a filename. It could be as simple as specifying a filename outside of the synced folder, and if it exists, then add the schedules. It could just be an empty file, you would just `touch ~/.schedule_bunch` on any machine you did want it to add schedules on, and by default it wouldn't run on anything else.

Could feasibly make that file a script instead, scheduling if it exits 0 and skipping otherwise. That would allow people to add their own criteria to test for, but would require everyone to know basic shell scripting.

I'm leaning toward the dummy file method, personally, but open to ideas.

**Reply by moffat** [reply]
Could the bunch installation itself generate some kind of ID that you could use for this purpose? Something that's repeatable[1] based on say the type and serial number of the computer, so that's truly unique? You could display that somewhere in the log or in the "About Bunch" window. The documentation would then instruct the user to find & use that ID to restrict a bunch to run only on a certain machine? Just an idea...

[1]: With this, I mean something that'd be the same if the user say uninstalls and reinstalls bunch.

**Reply by ttscoff** [reply]
For the time being I went ahead and added `schedule if` and `schedule unless` frontmatter keys. If a file is specified and it does exist (schedule if), scheduling is run, otherwise it's ignored. `schedule unless` is the inverse, only running if the file doesn't exist. I think it's a workable solution for now, but definitely still open to ideas.

**Reply by ttscoff** [reply]
In answer to your question, yes, I can get an identifier for the machine itself, either MAC address or UUID. I think it's easier to just use a trigger file, though. Same number of steps, really, but feels cleaner to me. Do you see a benefit to tying a Bunch to a specific address rather than just using a trigger file?

**Reply by moffat** [reply]
The ID solution seems less fiddly to me. I think the problem with the dummy file is that it's easy to imagine a situation where one dummy file wouldn't be enough. So the user might have to manage several of those.

**Reply by ttscoff** [reply]
I decided on both.

`schedule if` and `schedule unless` can take one or more UUIDs (comma-separated). If the format of the value isn't recognized as a UUID, it's assumed to be a file and scheduling depends on whether the file exists (if) or doesn't exist (unless)

UUIDs can be copied to the clipboard with the Copy UUID button in the preference pane.

I also added UUID recognition to the `startup` key. If `startup` is set to a UUID (or list of UUIDs), it will only run on the specified machines.

And, as a side note, I added a `folder.frontmatter` file to subfolders that lets you specify frontmatter for all Bunches in the folder (only if the Bunch itself doesn't define the affected key itself), including these scheduling and startup keys.

**Reply by ttscoff** [reply]
The update with these changes will be out in a few minutes for you to test :).

**Reply by moffat** [reply]
Hey, this latest version broke items being opened with the finder. Error says "Application Finder failed to launch" and a dialog pops up complaining "Application "Service Station Finder Extension" failed to launch and/or open..." and then the file I was trying to open. The bunch has:

```
Finder
- SOMEFILE_I_WANT_TO_OPEN
```

I fixed it by calling the actual programs that finder would call to open the files.

**Reply by ttscoff** [reply]
Creating an Issue for this one.
